File: 2020-11-13/Riddler Classic/riddler.py
# https://fivethirtyeight.com/features/can-you-snatch-defeat-from-the-jaws-of-victory/

## express

import random

def calcExpWinnings():
    values = [200] * 6 + [400] * 6 + [600] * 6 + [800] * 6 + [1000] * 6
    winnings = []
    for i in range(len(values)):
        winningsSoFar = sum(values[:i])
        if winningsSoFar <= 1000:
            dailyDoubleWinnings = 1000
        else:
            dailyDoubleWinnings = winningsSoFar
        winnings.append(winningsSoFar + dailyDoubleWinnings + sum(values[i+1:]))
    return sum(winnings) / len(winnings)

calcExpWinnings()

def estExpWinnings():
    results = []
    for j in range(10000000):
        values = [200] * 6 + [400] * 6 + [600] * 6 + [800] * 6 + [1000] * 6
        dd = random.sample(list(range(30)), 1)[0]
        winnings = []
        for i in range(30):
            winningsSoFar = sum(winnings)
            if i == dd:
                if winningsSoFar <= 1000:
                    winnings.append(1000)
                else:
                    winnings.append(winningsSoFar)
            else:
                winnings.append(values[i])
        results.append(sum(winnings))
    return sum(results) / len(results)

estExpWinnings()

def calcExpHuntWinnings():
    values = [600] * 30
    winnings = []
    for i in range(len(values)):
        winningsSoFar = sum(values[:i])
        if winningsSoFar <= 1000:
            dailyDoubleWinnings = 1000
        else:
            dailyDoubleWinnings = winningsSoFar
        winnings.append(winningsSoFar + dailyDoubleWinnings + sum(values[i+1:]))
    return sum(winnings) / len(winnings)

calcExpHuntWinnings()

def estExpHuntWinnings():
    results = []
    for j in range(10000000):
        values = [200] * 6 + [400] * 6 + [600] * 6 + [800] * 6 + [1000] * 6
        dd = random.sample(list(range(30)), 1)[0]
        winnings = []
        playOrder = random.sample(list(range(30)), 30)
        for i in playOrder:
            winningsSoFar = sum(winnings)
            if i == dd:
                if winningsSoFar <= 1000:
                    winnings.append(1000)
                else:
                    winnings.append(winningsSoFar)
            else:
                winnings.append(values[i])
        results.append(sum(winnings))
    return sum(results) / len(results)

estExpHuntWinnings()

def calcExpStratWinnings(): # not right
    values = [200] * 6 + [400] * 6 + [600] * 6 + [800] * 6 + [1000] * 6
    dailyDoubleDist = [2, 0, 3, 1, 2, 0, 286, 131, 202, 173, 214, 124, 703, 352, 581, 557, 533, 358, 866, 508, 795, 719, 763, 518, 602, 281, 497, 530, 448, 355]
    playOrder = list(np.argsort(np.array(dailyDoubleDist)))
    winnings = []
    for i in range(30):
        winningsSoFar = sum([values[x] for x in playOrder[:i]])
        if winningsSoFar <= 1000:
            dailyDoubleWinnings = 1000
        else:
            dailyDoubleWinnings = winningsSoFar
        winnings.append((winningsSoFar + dailyDoubleWinnings + sum([values[x] for x in playOrder[i:]])) * dailyDoubleDist[i] / sum(dailyDoubleDist))
    return sum(winnings)

calcExpStratWinnings()

def estExpStratWinnings():
    results = []
    for j in range(10000000):
        values = [200] * 6 + [400] * 6 + [600] * 6 + [800] * 6 + [1000] * 6
        dailyDoubleDist = [2, 0, 3, 1, 2, 0, 286, 131, 202, 173, 214, 124, 703, 352, 581, 557, 533, 358, 866, 508, 795, 719, 763, 518, 602, 281, 497, 530, 448, 355]
        dd = random.choices(list(range(30)), weights = dailyDoubleDist, k = 1)
        playOrder = list(np.argsort(np.array(dailyDoubleDist)))
        winnings = []
        for i in playOrder:
            winningsSoFar = sum(winnings)
            if i == dd:
                if winningsSoFar <= 1000:
                    winnings.append(1000)
                else:
                    winnings.append(winningsSoFar)
            else:
                winnings.append(values[i])
        results.append(sum(winnings))
    return sum(results) / len(results)

estExpStratWinnings()

## classic
from math import comb
import numpy as np
import matplotlib.pyplot as plt

def probOfWin(n,w):
    # n is the number of throws so far
    # w is the number of wins so far
    R = 101 - n # remaining throws
    N = 51 - w # needed to win
    if(w > n): # check if possible to win
        return np.nan
    elif(R < N):
        return 0
    elif(w >= 51):
        return 1.00
    else:
        pOfOutcome = []
        for r in range(N, R + 1):
            pOfOutcome.append(comb(R, r) / (2 ** R))
        return sum(pOfOutcome)

def probOfState(n,w):
    # n is the number of throws so far
    # w is the number of wins so far
    if(w > n): # check if possible to win
        return np.nan
    else:
        return comb(n, w) / (2 ** n)

probWinMatrix = np.zeros((102,102))
for W in range(102):
    for n in range(102):
        probWinMatrix[n][W] = probOfWin(W,n)
plt.imshow(probWinMatrix, cmap = 'hot', interpolation = 'nearest')

probStateMatrix = np.zeros((102,102))
for r in range(102):
    for c in range(102):
        if probWinMatrix[r][c] >= 0.99:
            probStateMatrix[r][c] = probOfState(c,r)
plt.imshow(probStateMatrix, cmap = 'hot', interpolation = 'nearest')

p99 = probStateMatrix.sum(axis = 0)
q99 = 1 - p99

probOf99Chance = []
for i in range(102):
    p = p99[i] * np.prod(q99[:i])
    probOf99Chance.append(p)

probOf99Chance
np.sum(probOf99Chance)

probWinMatrix

File: 2020-11-13/Riddler Classic/riddler.R
# https://fivethirtyeight.com/features/can-you-snatch-defeat-from-the-jaws-of-victory/

calcPOfWinning <- function(n, h, N){
  # n is the number of throws so far
  # h is the number of heads so far
  # N is the total number of throws
  # H is the heads needed to win
  H <- ceiling((N + 1) / 2)

  # R is the heads needed left to win
  R <- H - h
  # r is the remaining throws
  r <- N - n
  q <- 0
  if(h > n){
    q <- 0
  } else if(R > r){
    q <- 0
  } else {
    for(i in R:r){
      q <- q + (choose(r, i) / 2 ^ r)
    }
  }
  q
}

calcPOfState <- function(n, h){
  # n is the number of throws so far
  # h is the number of heads so far
  if(h < n){
    q <- 0
  } else {
    q <- choose(n, h) / 2 ^ n
  }
}

fillPWinMatrix <- function(N){
  pMat <- matrix(NA, ncol = N, nrow = N)
  for(i in 1:N){
    for(j in 1:N){
      pMat[i,j] <- calcPOfWinning(i-1, j-1, N)
    }
  }
  pMat
}

transformPWinMatrix <- function(m, p = .99){
  for(i in dim(m)[1]){
    for(j in dim(m)[2]){
      if(m[i,j] < p | m[i,j] == 1){
        m[i,j] <- 0
      } else {
        m[i,j] <- 1 - m[i,j]
      }
    }
  }
  m
}

fillPStateMatrix <- function(N){
  pMat <- matrix(NA, ncol = N, nrow = N)
  for(i in 1:N){
    for(j in 1:N){
      pMat[i,j] <- calcPOfState(i-1, j-1)
    }
  }
  pMat
}


playGame <- function(N){
  H <- ceiling((N + 1) / 2)
  flips <- sample(c(0,1), N, replace = T)
  n <- 0
  h <- 0
  p <- 0
  for(f in flips){
    q <- calcPOfWinning(n, h, N)
    if(q > p){
      p <- q
    }
    h <- h + f
    n <- n + 1
  }
  r <- p > .99 & h < H
  r
}

playGameRepeatedly <- function(N){
  x <- replicate(10000, playGame(N))
  y <- mean(x)
  y
}

playGameRepeatedly(1001)
